src/System.IO.Pipelines/Pipe.cs

```diff
@@ -440,28 +437,26 @@ void IPipeReader.Advance(ReadCursor consumed, ReadCursor examined)
                 _readHead.Start = consumed.Index;
             }
 
-            bool resumeWriter;
+            var currentLength = Interlocked.Add(ref _length, -consumedBytes);
 
             // Reading commit head shared with writer
+            bool consumedEverything;
             lock (_sync)
             {
-                _length -= consumedBytes;
-                resumeWriter = _length < _maximumSizeLow;
-
-                // Change the state from observed -> not cancelled. We only want to reset the cancelled state if it was observed
-                Interlocked.CompareExchange(ref _cancelledState, CancelledState.NotCancelled, CancelledState.CancellationObserved);
+                consumedEverything = examined.Segment == _commitHead &&
+                                     examined.Index == _commitHeadIndex &&
+                                     !_writerCompletion.IsCompleted;
+            }
 
-                var consumedEverything = examined.Segment == _commitHead &&
-                                         examined.Index == _commitHeadIndex &&
-                                         !_writerCompletion.IsCompleted;
+            // Change the state from observed -> not cancelled. We only want to reset the cancelled state if it was observed
+            var cancelledState = Interlocked.CompareExchange(ref _cancelledState, CancelledState.NotCancelled, CancelledState.CancellationObserved);
 
-                // We reset the awaitable to not completed if
-                // 1. We've consumed everything the producer produced so far
-                // 2. Cancellation wasn't requested
-                if (consumedEverything && _cancelledState != CancelledState.CancellationRequested)
-                {
-                    Reset(ref _readerCallback);
-                }
+            // We reset the awaitable to not completed if
+            // 1. We've consumed everything the producer produced so far
+            // 2. Cancellation wasn't requested
+            if (consumedEverything && cancelledState != CancelledState.CancellationRequested)
+            {
+                Reset(ref _readerCallback);
```

Comment:
There's still a chance that CancelPendingRead() was called between the CompareExchange() and the call to Reset. This would cause the subsequent call to ReadAsync() to stall even though in the Pipe is in the CancellationObserved state. I think it could be changed to the following:

```c#
// We reset the awaitable to not completed if we've consumed everything the producer produced so far
if (consumedEverything)
{
    Reset(ref _readerCallback);
}

// Change the state from observed -> not cancelled. We only want to reset the cancelled state if it was observed
var cancelledState = Interlocked.CompareExchange(ref _cancelledState, CancelledState.NotCancelled, CancelledState.CancellationObserved);

if (consumedEverything && cancelledState == CancelledState.CancellationRequested)
{
    // Rollback the awaitable reset if a cancellation was requested
    Resume(_readerScheduler, ref _readerCallback);
}
```